index.md

```diff
@@ -1453,14 +1448,415 @@ def check(condition, message):
+- [StackOverflow: About the changing id of an immutable string]
+- [StackOverflow: When does Python allocate new memory for identical strings?]
+
+### Unsupported assignment operation (E1137) {#E1137}
+
+This error occurs when we assign something to an object which does not support assignment (i.e. an object which does not define the `__setitem__` method).
+
+~~~~ {include="E1137_unsupported_assignment_operation"}
+~~~~
+
+While we cannot mutate an immutable object, often, we can create a new object instead.
+
+```python
+my_string = "Hello World!"
+my_string = my_string[:6] + "Universe!"
+print(my_string)  # Prints 'Hello Universe!'
+```
+
+### Expression not assigned (W0106) {#W0106}
+
+This error occurs when an expression that is not a function call is not assigned to a variable. Typically, this indicates that we were indenting to do something else.
+
+~~~~ {include="W0106_expression_not_assigned"}
+~~~~
+
+Corrected version:
+
+```python
+lst = [1, 2, 3]
+lst.append(4)
+print("Appended 4 to my list!")
+```
+
+### Invalid length returned (E0303) {#E0303}
+
+This error occurs when the `__len__` special method returns something other than a non-negative integer.
+
+~~~~ {include="E0303_invalid_length_returned"}
+~~~~
+
+Corrected version:
+
+```python
+class Company:
+
+    def __init__(self, employees: List[str]) -> None:
+        self._employees = employees
+
+    def __len__(self) -> int:
+        return len(self._employees)
+```
+
+### Abstract method (W0223) {#W0223}
+
+This error occurs when an abstract method (i.e. a method with a `raise NotImplementedError` statement) is not overridden inside a concrete class.
+
+~~~~ {include="W0223_abstract_method"}
+~~~~
+
+Corrected version:
+
+```python
+class Cat(Animal):
+
+    def make_sound(self) -> str:
+        return 'Miew...'
+```
+
+### Arguments differ (W0221) {#W0221}
+
+This error occurs when a method takes a different number of arguments than the interface that it implements or the method that it overrides.
+
+~~~~ {include="W0221_arguments_differ"}
+~~~~
+
+Corrected version:
+
+```python
+class Dog(Animal):
+    """Class representing a dog."""
+
+    def make_sound(self, mood: str) -> None:
+        if mood == 'happy':
+            print("Woof Woof!")
+        elif mood == 'angry':
+            print("Grrrrrrr!!")
+```
+
+### Unexpected keyword arg (E1123) {#E1123}
+
+This error occurs when a function call passes a keyword argument which does not match the signature of the function being called.
+
+~~~~ {include="E1123_unexpected_keyword_arg"}
+~~~~
+
+Corrected version:
+
+```python
+print_greeting(name="Arthur")
+```
+
+### Redefined argument from local (R1704) {#R1704}
+
+This error occurs when a local name is redefining the name of a parameter. We should avoid reusing the name of a parameter in binding operations such as for iteration, with statement assignment, and exception handler assignment, as this needlessly obfuscates the code.
```

Comment:
I think you can remove the second sentence (the terminology isn't helpful here), and also reference the other name shadowing-type checks.